source/java/src/org/lucee/extension/cache/eh/util/EHCacheClassLoader.java

```diff
@@ -0,0 +1,135 @@
+package org.lucee.extension.cache.eh.util;
+
+import java.lang.reflect.Method;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+
+import lucee.loader.engine.CFMLEngine;
+import lucee.loader.engine.CFMLEngineFactory;
+import lucee.runtime.config.Config;
+
+import lucee.commons.io.log.Log;
+
+
+/**
+ * A special class loader that first attempts to load class from the Lucee Environment,
+ * but that can fail when RMI classes are requested via a spawned thread, so when
+ * the class cannot be located in the Lucee Environment, it searches all the Felix
+ * bundles for the requested class. This will prevent issues with issues trying
+ * to load classes like net.sf.ehcache.distribution.RMICachePeer_Stub. 
+ *
+ * @author Dan G. Switzer, II
+ * @version 1.0
+ */
+public class EHCacheClassLoader extends ClassLoader {
+	private Config config;
+	private ClassLoader LuceeEnvClassLoader;
+
+	public EHCacheClassLoader(Config config) {
+		this.config = config;
+
+		try {
+			Method m = config.getClass().getMethod("getClassLoaderEnv", new Class[0]);
+			this.LuceeEnvClassLoader = (ClassLoader) m.invoke(config, new Object[0]);
+		}
+		catch (Exception e) {
+			// do nothing
+		}
+	}
+
+	public CFMLEngine getCFMLEngine(){
```

Comment:
i don't get this method, "CFMLEngineFactory.getInstance()" already gives you the CFMLEngine, you even use CFMLEngine on line 43 to get CFMLEngine.

Reply:
My memory is a little foggy on this one, but I'm pretty sure the issue is that `CFMLEngineFactory.getInstance()` doesn't always return the expected CFMLEngine instance, such as when running asynchronously. The only way I could find to get it to return the instance configured the way it was needed was to call `getEngine()` and pass in the correct Config instance.

I've added a comment to the code to explain this. I apologize for not doing this in the first place. Bad programmer! Bad! :)

Reply:
In MY opinion that is impossible,  because the engine get loaded by loader and within one classloader it is always only one engine ever, so it cannot be that we see a different engine within one classloader. it also never exist 2 engine instances at the same time. If that would be the case, we would have a SERIOUS issue that must be addressed FIRST thing. We cannot have code not trusting the basis of the environment and does things like this. 
If that really is an issue, can you please raise a ticket and we will address it first thing. Best with some test code.

Reply:
@michaeloffner,

You are correct, it does not appear that the `public CFMLEngine getCFMLEngine(){` method is necessary, so I've removed it.

Sorry it took so long to test this. I needed to rebuild some VMs because I thought the problem was related to the asynchronous operations that happened when caching was being expired in the background. So I needed to test in the environment that I was using when I was originally working on this code.